[GSoC] A few leading term methods changed

0sidharth opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #20365
Fixes #21031
Fixes #21606

#### Brief description of what is fixed or changed
`Pow`, `exp`, `factorial` and `gamma` `_eval_as_leading_term` now raise `PoleError` when the argument is unbounded, instead of returning the original expression itself.
Add `LambertW._eval_as_leading_term` method
Fixed `besselj` and `bessely` leading term methods
Added `erfi` leading term method
Added `SingularityFunction` leading term method and power series expansion
Added `fresnels` and `fresnelc` leading term methods

#### Other comments
The previous implementation of `bound` parameter for `aseries` was buggy and misleading, and has broken after these changes. After thinking about what the `bound` parameter should actually represent, it should be coded properly. Consequently, the corresponding tests have been commented out for now.

#7529 and #7552 may be closed now as all the code in those PRs will be present in master
(For #7552, the code is there in #17303, so only one of the two need to remain open)

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * `exp`, `factorial` and `gamma` `._eval_as_leading_term` now raises `PoleError` when the argument is unbounded
  * Add `LambertW`, `erfi`, `fresnels`, `fresnelc` and `SingularityFunction` leading term methods
  * Add `SingularityFunction` power series expansion
  * Fixed `besselj` and `bessely` leading term methods
* series
  * Limit code has better handling of `sign`
* core
  * `Pow._eval_as_leading_term` now raises `PoleError` when the argument is unbounded
  * `leadterm` now takes `logx` argument
<!-- END RELEASE NOTES -->

0sidharth committed:
Add LambertW._eval_as_leading_term method

The try except block in the limit code for meromorphic function
is removed after this addition, as LambertW was the only instance
throwing an error and requiring the try except block

0sidharth committed:
Fixed besselj and bessely leading term methods

jksuom commented on sympy/core/expr.py:
```diff
@@ -2962,10 +2962,8 @@ def series(self, x=None, x0=0, n=6, dir="+", logx=None, cdir=0):
-            if n is None:
-                return (si.subs(x, sgn/x) for si in s)
-            return s.subs(x, sgn/x)
+            s = self.subs(x, sgn*x).aseries(x, n).subs(x, sgn*x)
```
Many functions have proper power series expansions at infinity; rational functions, for example, have expansions everywhere. Those should usually be created by `nseries`. Asymptotic series may be located anywhere in the extended complex plane, not necessarily at infinity. `aseries` should be called only when a power series does not exist.

0sidharth commented on sympy/core/add.py:
```diff
@@ -1009,7 +1009,7 @@ def _eval_as_leading_term(self, x, cdir=0):
             new_expr = new_expr.trigsimp().cancel()
             is_zero = new_expr.is_zero
-        if is_zero is True:
+        if is_zero is True and not Order(0, x).contains(min):
```
`min` is sometimes `Order(0, x)` after the changes in this PR, so `min.getn()` throws an error

jksuom replied:
Does this mean that `min` is `Order(0, x)`?  That could be tested by something like `min.expr == 0` without creating a new `Order(0, x)` and invoking `contains` (which is not cheap in general).

jksuom commented on sympy/functions/elementary/exponential.py:
```diff
@@ -1173,6 +1173,14 @@ def _eval_is_algebraic(self):
 
+    def _eval_as_leading_term(self, x, cdir=0):
+        if len(self.args) == 1:
+            arg0 = self.args[0]
+            arg = self.args[0].subs(x, 0).cancel()
```
```suggestion
            arg = self.args[0]
            arg0 = arg.subs(x, 0).cancel()
```
Also `arg0` and `arg` swapped on the next lines.

jksuom commented on sympy/core/expr.py:
```diff
@@ -2961,9 +2961,16 @@ def series(self, x=None, x0=0, n=6, dir="+", logx=None, cdir=0):
             sgn = 1 if x0 is S.Infinity else -1
-            s = self.subs(x, sgn*x).aseries(x, n).subs(x, sgn*x)
-            return s
+            try: #Check if a power series form exists at oo
```
A space after # looks better.

jksuom commented on sympy/series/tests/test_series.py:
```diff
@@ -29,6 +29,12 @@ def test_exp2():
     assert e1 == e2
 
 
+def test_asymptotic_series():
```
This looks like the same test as above in test_aseries. Maybe that alone would suffice?

0sidharth replied:
This test is needed as the call here is to `series(x, oo)`, which raises `PoleError` on master (which doesn't call) `aseries`. Perhaps the test case in test_aseries can be removed

jksuom replied:
Ok.

0sidharth committed:
Added erfi leading term method

0sidharth committed:
LambertW leadterm: variable name changes

0sidharth committed:
remove_abs in limit now handles sign and renamed to set_sign

0sidharth committed:
zero checking instead of Order(0, x).contains

jksuom commented on sympy/functions/special/error_functions.py:
```diff
@@ -613,6 +613,15 @@ def _eval_expand_func(self, **hints):
 
+    def _eval_as_leading_term(self, x, cdir=0):
+        arg = self.args[0].as_leading_term(x)
+        arg0 = arg.subs(x, 0).cancel()
```
Is `cancel` necessary?

0sidharth committed:
Edit leading term methods to now take logx parameter (default None)

0sidharth committed:
Added logx parameter to a few _eval_as_leading_term calls
log._eval_as_leading_term now raises about nan
Expr.dir now raises if nan is encountered
Rename set_sign to set_signs in limits.py

0sidharth committed:
Removed unnecessary `cancel`

0sidharth committed:
Temporarily XFAILed a couple of tests

0sidharth committed:
Merge branch 'master' into lambert_leading_term

0sidharth committed:
Flake8 changes

0sidharth committed:
Added AttributeError catching in limits.py

0sidharth committed:
Added SingulairtyFunction leading term method

0sidharth committed:
Remove comment

0sidharth committed:
Added test case for issue 20365

0sidharth committed:
Pow._eval_as_leading term now raises when argument is unbounded
Previously, Pow and exp _eval_as_leading_term methods returned the
original expression itself when the argument was unbounded. Now, they
raise a PoleError (example - exp(1/x).as_leading_term(x) raises an error
and the code has accordingly been changed to account for this.

Due to the raising of an exception, the Order.__new__ code had to be
changed, as it relied on finding the leading term of the argument.
Now, it calculates the leading term of the args of its argument, and
depending on the type of argument, performs some simplifcation and
returns the answer.

fresnels and fresnelcs _eval_as_leading_term methods added

0sidharth committed:
Added bottom_up for normalizing expressions in gruntz
Added PoleError catching in exp._eval_nseries
mrv_leadterm now handles logx substitution
Added test case of #21031
Rewrote failing tests

0sidharth commented on sympy/core/add.py:
```diff
@@ -1009,16 +1009,16 @@ def _eval_as_leading_term(self, x, cdir=0):
             new_expr = new_expr.trigsimp().cancel()
             is_zero = new_expr.is_zero
-        if is_zero is True:
+        if is_zero is True and min.expr != 0:
```
`min.expr != 0` check is no longer needed

0sidharth commented on sympy/integrals/integrals.py:
```diff
@@ -1149,19 +1149,19 @@ def _eval_lseries(self, x, logx, cdir=0):
                 break
         terms, order = expr.function.nseries(
-            x=symb, n=n, logx=logx).as_coeff_add(Order)
+            x=symb, n=n).as_coeff_add(Order)
```
This change is also not needed, `logx` parameter can be kept, though it won't make a difference

0sidharth committed:
Add SingularityFunction _eval_nseries method
Fixed typos and bugs in SingularityFunction leading term
Remove unneeded changes

jksuom commented on sympy/functions/special/error_functions.py:
```diff
@@ -613,6 +613,15 @@ def _eval_expand_func(self, **hints):
+        if x in arg.free_symbols and arg0.is_zero:
+            return 2*arg/sqrt(pi)
+        else:
+            return self.func(arg)
```
Shouldn't this have arg0?

jksuom commented on sympy/functions/special/singularity_functions.py:
```diff
@@ -210,5 +210,31 @@ def _eval_rewrite_as_Heaviside(self, *args, **kwargs):
             return (x - a)**n*Heaviside(x - a)
 
+    def _eval_as_leading_term(self, x, logx=None, cdir=0):
+        z = self.args[0]
```
How about something like `z, a, n = self.args`?

jksuom commented on sympy/core/power.py:
```diff
@@ -1703,36 +1708,29 @@ def mul(d1, d2):
-            arg_1 = arg.as_leading_term(x)
-            if Order(x, x).contains(arg_1):
+            arg = e.as_leading_term(x, logx=logx)
+            if Order(x, x).contains(arg):
```
This is probably intended to mean that `arg.subs(x, 0).is_zero`. However, it does not cover all such cases:
```
In [18]: Order(x, x).contains(sqrt(x))                                          
Out[18]: False
```

jksuom commented on sympy/core/power.py:
```diff
@@ -1703,36 +1708,29 @@ def mul(d1, d2):
-            # instead of None.                                 #
-            ####################################################
-            return S.Exp1**arg
+            if Order(1, x).contains(arg):
```
This could probably be `arg.subs(x, 0).is_finite`. (`arg` should be well-behaved when `as_leading_term` does not raise.)

0sidharth committed:
Merge branch 'master' into lambert_leading_term

0sidharth committed:
Gamma and factorial leading term methods now raise PoleError if arg is unbounded
They previously returned the original function itself

Contains checks replaced by is_zero and not is_infinite checks

jksuom commented on sympy/functions/special/tests/test_error_functions.py:
```diff
@@ -201,7 +201,7 @@ def test_erfi():
     assert erfi(x*y).as_leading_term(y) == 2*x*y/sqrt(pi)
     assert (erfi(x*y)/erfi(y)).as_leading_term(y) == x
-    assert erfi(1/x).as_leading_term(x) == erfi(1/x)
+    assert erfi(1/x).as_leading_term(x) == 1/(x*sqrt(-1/x**2))
```
Could this be expressed in terms of `sign(x)`?

0sidharth replied:
That would need the assumption that x is real. Should that be made?

jksuom replied:
Actually, it seems that `erfi(x)` will be of the order of `exp(x**2)` as `x` tends to infinity so there will be no leading term for real `x`. For pure imaginary `x`, it is bounded (`-I*erf(I*x)`), but that should not considered here. So it looks like the call should raise.

0sidharth committed:
Better handling for zoo in erf functions
Limits code finding leading term of non-meromorphic functions inverts cdir

jksuom commented on sympy/functions/special/tests/test_error_functions.py:
```diff
@@ -48,7 +48,7 @@ def test_erf():
     assert erf(x*y).as_leading_term(y) == 2*x*y/sqrt(pi)
     assert (erf(x*y)/erf(y)).as_leading_term(y) == x
-    assert erf(1/x).as_leading_term(x) == S.One
+    assert erf(1/x).as_leading_term(x) == erf(1/x)
```
`erf` has finite limits `+/-1` at `+/-oo` (unlike `erfi`) so this test need not be changed.

0sidharth committed:
erf(1/x).as_leading_term(x) == 1 # instead of erf(1/x)

jksuom commented on sympy/functions/special/error_functions.py:
```diff
@@ -233,12 +233,12 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         if x in arg.free_symbols and arg0.is_zero:
             return 2*arg/sqrt(pi)
-        elif arg0.is_finite:
+        elif arg0.is_extended_real:
```
`arg0` may be finite but not real.

jksuom commented on sympy/functions/special/error_functions.py:
```diff
@@ -233,12 +233,12 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         if x in arg.free_symbols and arg0.is_zero:
             return 2*arg/sqrt(pi)
-        elif arg0.is_finite:
+        elif arg0.is_number:
```
We might have `arg = x + y` and `arg0 = y`. Then the result should be `erf(y)`.

jksuom replied:
The test could possibly be `if not arg0.is_infinite`.

jksuom commented on sympy/functions/special/error_functions.py:
```diff
@@ -233,12 +233,12 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
             return self.func(arg0)
-        elif arg0 is S.ComplexInfinity:
-            return self.func(arg)
         return -S.One if cdir == -1 else S.One
```
`cdir` should be in `limit(arg, x, 0, cdir=cdir)`.

jksuom commented on sympy/functions/special/error_functions.py:
```diff
@@ -233,12 +233,12 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         arg = self.args[0].as_leading_term(x)
```
We should probably put `logx` and `cdir` here as well.

0sidharth committed:
Code review changes

jksuom commented on sympy/integrals/tests/test_integrals.py:
```diff
@@ -1507,6 +1507,7 @@ def test_issue_14782_slow():
+@XFAIL
 def test_issue_12081():
     f = x**(Rational(-3, 2))*exp(-x)
     assert integrate(f, [x, 0, oo]) is oo
```
Why does this fail?

jksuom commented on sympy/series/order.py:
```diff
@@ -221,7 +225,37 @@ def __new__(cls, expr, *args, **kwargs):
+                                    order = Order(lt, *pts)
+                                orders.append(order)
+                            if expr.is_Add:
+                                new_expr = Order(Add(*orders), *pts)
```
How shall `new_expr` be used?

jksuom commented on sympy/series/tests/test_residues.py:
```diff
@@ -80,7 +80,7 @@ def test_issue_14037():
 def test_issue_21176():
     f = x**2*cot(pi*x)/(x**4 + 1)
     assert residue(f, x, -sqrt(2)/2 - sqrt(2)*I/2).together() == \
```
It looks like `cancel().together(deep=True)` would give a better looking result.

jksuom commented on sympy/series/tests/test_series.py:
```diff
@@ -328,12 +328,12 @@ def test_issue_20697():
-        (36/(-36*sqrt(5) - 80) + 16*sqrt(5)/(-36*sqrt(5) - 80))/(x - 1/(S.Half\
-        + sqrt(5)/2)) - 1220*sqrt(5)/(-6100*sqrt(5) - 13640) - 2728\
-        /(-6100*sqrt(5) - 13640) + O(x - 2/(1 + sqrt(5)), (x, 2/(1 + sqrt(5))))
+    assert (1/(1 - x - x**2)).series(x, 1/fi, 1).factor() == \
```
The result could probably be improved. I tried with `radsimp` but that raises
```
ValueError: expecting a Symbol but got _Dummy_628 - sqrt(5)/2 + 1/2 - (1/2 - sqrt(5)/2)
```
It looks like this would come from this line: https://github.com/sympy/sympy/blob/1a9577f9220fc24372ad4bd6d97059759a6b665c/sympy/series/order.py#L193
because
```
(Pdb) p rs
{_Dummy_628 - sqrt(5)/2 + 1/2 - (1/2 - sqrt(5)/2): x - sqrt(5)/2 + 1/2}
```
Obviously, some simplification is necessary for `v - point[0]`.

0sidharth committed:
Added simplification in Add._eval_as_leading_term for #21334
Fixed failing test in integrals
Change name of set_sign to set_signs and test_abs_sign to test_set_signs
Fixed a bug in Order.__new__ in the case of Addition of multiple Orders
Code Review changes

jksuom commented:
This is ready to go. Thanks!
